### PR TITLE
Properly mark IOperation nodes as Implicit.

### LINKED ISSUE #7397: An empty VB method should have no operations in that block.
Consider an empty VB method:

``` vb.net
Sub foo()
End Sub
```

If I register an OperationBlockAction and inspect the Operations in that block, I'd expect it to have no operations.

Instead, it has one LabelStatement operation for an Exit label. This is implementation detail that shouldn't show up in the operation tree. C# doesn't have this issue.

### LINKED ISSUE #22679: [IOperation] Should IUsingStatement's block be implicit
See https://github.com/dotnet/roslyn/pull/22630#discussion_r144369564. Crux of the argument is that the IBlockStatement under an IUsingStatement in VB uses the same syntax as the IUsingStatement, but isn't currently implicit. Should that block be considered implicit?

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #22679. Fixes #7397. Related to #22029.

Tracking VSO bug :https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems/edit/510300

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation, @cston Please review.

**Comment by AlekseyTs** [comment]
CC @jinujoseph For ask mode approval.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 1]
```diff
@@ -3118,9 +3120,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             If targetTypeIsValid Then
                 initialValue = ApplyImplicitConversion(initialValue.Syntax, targetType, initialValue, diagnostics)
                 limit = ApplyImplicitConversion(limit.Syntax, targetType, limit, diagnostics)
+                Dim stepValueBeforeConversion = stepValue
                 stepValue = ApplyConversion(stepValue.Syntax, targetType, stepValue,
                                             isExplicit:=forStatement.StepClause Is Nothing,
                                             diagnostics:=diagnostics)
+
+                If stepValue IsNot stepValueBeforeConversion AndAlso stepValue.Kind = BoundKind.Conversion AndAlso
+                   forStatement.StepClause Is Nothing Then
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Is it sufficient to check `forStatement.StepClause Is Nothing`, or perhaps `stepValueBeforeConversion.WasCompilerGenerated()`?

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 2]
```diff
@@ -4803,7 +4811,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Next
             End If
 
-            Dim block = Me.BindBlock(node, node.Statements, diagnostics)
+            Dim block = Me.BindBlock(node, node.Statements, diagnostics).MakeCompilerGenerated()
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: It looks like most `BoundBlock` instances are now marked compiler generated, but method body blocks are not. What is the difference with method body blocks?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 3]
```diff
@@ -1216,10 +1216,16 @@ Namespace Microsoft.CodeAnalysis.Semantics
             Dim syntax As SyntaxNode = boundReturnStatement.Syntax
             Dim type As ITypeSymbol = Nothing
             Dim constantValue As [Optional](Of Object) = New [Optional](Of Object)()
-            Dim isImplicit As Boolean = boundReturnStatement.WasCompilerGenerated
+            Dim isImplicit As Boolean = boundReturnStatement.WasCompilerGenerated OrElse IsEndSubOrFunctionStatement(syntax)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: When is `boundReturnStatement.WasCompilerGenerated = False` and `IsEndSubOrFunctionStatement(syntax) = True`?

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 4]
```diff
@@ -2710,9 +2712,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Dim alternative As BoundStatement = Nothing
 
             condition = BindBooleanExpression(node.Condition, diagnostics)
-            consequence = BindBlock(node, node.Statements, diagnostics)
+            consequence = BindBlock(node, node.Statements, diagnostics).MakeCompilerGenerated()
             If node.ElseClause IsNot Nothing Then
-                alternative = BindBlock(node.ElseClause, node.ElseClause.Statements, diagnostics)
+                alternative = BindBlock(node.ElseClause, node.ElseClause.Statements, diagnostics).MakeCompilerGenerated()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: >.MakeCompilerGenerated() [](start = 97, length = 24)

I am planning to revert this change in the upcoming commit.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 5]
```diff
@@ -2728,18 +2730,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                 conditions.Add(BindBooleanExpression(elseIfBlock.ElseIfStatement.Condition, diagnostics))
-                blocks.Add(BindBlock(elseIfBlock, elseIfBlock.Statements, diagnostics))
+                blocks.Add(BindBlock(elseIfBlock, elseIfBlock.Statements, diagnostics).MakeCompilerGenerated())
             Next
 
             Dim currentAlternative As BoundStatement = Nothing
 
             If node.ElseBlock IsNot Nothing Then
-                currentAlternative = BindBlock(node.ElseBlock, node.ElseBlock.Statements, diagnostics)
+                currentAlternative = BindBlock(node.ElseBlock, node.ElseBlock.Statements, diagnostics).MakeCompilerGenerated()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: >.MakeCompilerGenerated() [](start = 102, length = 24)

I am planning to revert this change in the upcoming commit.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: > What is the difference with method body blocks?

Their associated syntax doesn't conflict with other IOperation nodes. Also, GetOperation API returns only explicit IOperation nodes.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > When is boundReturnStatement.WasCompilerGenerated = False and IsEndSubOrFunctionStatement(syntax) = True?

See unit-test IBlockStatement_SubMethodBlock for example.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: I hesitate doing this on arbitrary nodes because it is not allowed to mark nodes as compiler generated after this state was checked by someone.

**Comment by natidea** [comment]
Approved to merge via [510300](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/510300)

**Comment by AlekseyTs** [comment]
Ping @dotnet/roslyn-compiler, @cston Please review 15.5 fixes for several IOperation issues. They are approved by ship-room and should be merged ASAP. Thanks.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 6]
```diff
@@ -383,7 +383,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     locals = localBuilder.ToImmutableAndFree()
                 End If
 
-                statements.Add(New BoundReturnStatement(methodBlock.EndBlockStatement, New BoundLocal(methodBlock.EndBlockStatement, localForFunctionValue, isLValue:=False, type:=localForFunctionValue.Type), Nothing, Nothing))
+                statements.Add(New BoundReturnStatement(methodBlock.EndBlockStatement,
```
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Should we mark the return statement with end block syntax as implicit, so that empty sub methods have no explicit operations?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb (hunk 3 above)
Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Ah, ignore my previous comment, I see you are handling it here.

**Review by mavasani — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: > Should we mark the return statement with end block syntax as implicit, so that empty sub methods have no explicit operations?

This flag also affects how we emit PDB sequence points. Marking nodes here would affect debugging experience.

**Review by OmarTawfik — APPROVED** [review]
LGTM
